### PR TITLE
secure: grpc on https

### PR DESCRIPTION
**Pull request by noxiouz**

### CODE REVIEW

File: insonmnia/miner/builder.go [hunk 1]
```diff
@@ -139,7 +144,26 @@ func (b *MinerBuilder) Build() (miner *Miner, err error) {
...
+		if err != nil {
+			return nil, err
+		}
+		var cert = tls.Certificate{
+			Certificate: [][]byte{certX509.Raw},
+			PrivateKey:  priv,
+		}
+		servOpts = append(servOpts, grpc.Creds(util.NewTLS(&tls.Config{Certificates: []tls.Certificate{cert}})))
+	}
+	grpcServer := grpc.NewServer(servOpts...)
```
Comment (3Hren) [inline insonmnia/miner/builder.go]: Seems like there are options that are duplicated, like `RPCCompressor` etc here and in the `utils.MakeGrpcServer`. Maybe reuse?

File: util/grpcsecure.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		if err != nil {
+			return nil, err
+		}
+		return EthAuthInfo{TLS: authInfo, Wallet: wallet}, nil
+	default:
+		return nil, fmt.Errorf("unsupported AuthInfo %s %T", authInfo.AuthType(), authInfo)
+	}
+}
+
+// NewTLS ...
```
Comment (3Hren) [inline util/grpcsecure.go]: So mysterious...

File: util/certs.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		return nil, nil, fmt.Errorf("failed to sign public key: %v", err)
+	}
+	issuerCommonName.Write(signature.Serialize())
+
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(100),
+		Subject: pkix.Name{
+			CommonName: base32.StdEncoding.EncodeToString(issuerCommonName.Bytes()),
+		},
+		NotBefore: time.Now().Add(-time.Hour * 24 * 7),
```
Comment (3Hren) [inline util/certs.go]: Should we be able to configure this?

Comment (noxiouz) [reply util/certs.go]: Makes more sense to configure NotAfter :)

File: insonmnia/miner/builder.go (hunk 1 above)
Comment (noxiouz) [reply insonmnia/miner/builder.go]: Gonna reuse, yes. Good idea.

File: insonmnia/miner/builder.go [hunk 4]
```diff
@@ -139,7 +144,26 @@ func (b *MinerBuilder) Build() (miner *Miner, err error) {
...
-	grpcServer := grpc.NewServer(grpc.RPCCompressor(grpc.NewGZIPCompressor()), grpc.RPCDecompressor(grpc.NewGZIPDecompressor()))
+	var servOpts = []grpc.ServerOption{grpc.RPCCompressor(grpc.NewGZIPCompressor()), grpc.RPCDecompressor(grpc.NewGZIPDecompressor())}
+	if os.Getenv("GRPC_INSECURE") == "" {
+		// TODO: read from config
+		ethKey, err := ethcrypto.GenerateKey()
+		if err != nil {
+			cancel()
+			return nil, err
+		}
+		// Generate Certificate with signed publick key
```
Comment (3Hren) [inline insonmnia/miner/builder.go]: Typo: `publick`.

File: insonmnia/miner/builder.go [hunk 5]
```diff
@@ -139,8 +146,34 @@ func (b *MinerBuilder) Build() (miner *Miner, err error) {
...
+		// The rotator will be stopped by ctx
+		certRotator, TLSConf, err = util.NewHitlessCertRotator(ctx, ethKey)
+		if err != nil {
+			return nil, err
+		}
+		creds = util.NewTLS(TLSConf)
+	}
+	grpcServer := util.MakeGrpcServer(creds)
 
+	fmt.Println("SSSS", b.cfg.HubResources())
```
Comment (sshaman1101) [inline insonmnia/miner/builder.go]: "SSSS" is like snakes do? But this is not a Python :)

Comment (noxiouz) [reply insonmnia/miner/builder.go]: Damn it) now you know, that I use println debugging.

**Comment by noxiouz** [comment]
@3Hren ready to listen to ur comments

**Pull request merged**